File: data structures and algorithms/java/08_043.java
/**
 * Finds maximum diagonal sum in a square matrix
 */
public class Solution {
   public int maxDiagonalSum(int[][] matrix) {
      if (matrix == null || matrix.length == 0 || 
          matrix[0].length == 0) {
         return 0;
      }
      
      int n = matrix.length;
      int maxSum = Integer.MIN_VALUE;
      
      // Check diagonals starting from top row
      for (int col = 0; col < n; col++) {
         int diagonalSum = 0;
         int r = 0, c = col;
         
         while (r < n && c < n) {
            diagonalSum += matrix[r][c];
            r++;
            c++;
         }
         
         maxSum = Math.max(maxSum, diagonalSum);
      }
      
      // Check diagonals starting from leftmost column
      for (int row = 1; row < n; row++) {
         int diagonalSum = 0;
         int r = row, c = 0;
         
         while (r < n && c < n) {
            diagonalSum += matrix[r][c];
            r++;
            c++;
         }
         
         maxSum = Math.max(maxSum, diagonalSum);
      }
      
      return maxSum;
   }
}

File: data structures and algorithms/python/08_043.py
def max_diagonal_sum(matrix):

    if not matrix or not matrix[0]:
        return 0
    
    n = len(matrix)
    max_sum = float('-inf')
    
    # Check diagonals starting from top row
    for col in range(n):
        diagonal_sum = 0
        r, c = 0, col
        
        while r < n and c < n:
            diagonal_sum += matrix[r][c]
            r += 1
            c += 1
        
        max_sum = max(max_sum, diagonal_sum)
    
    # Check diagonals starting from leftmost column
    # Start from 1 to avoid recounting (0,0)
    for row in range(1, n):  
        diagonal_sum = 0
        r, c = row, 0
        
        while r < n and c < n:
            diagonal_sum += matrix[r][c]
            r += 1
            c += 1
        
        max_sum = max(max_sum, diagonal_sum)
    
    return max_sum